File: cannabis_management/cannabis_management/custom/stock_entry.py
import frappe
from frappe import _
from frappe.utils import flt


def before_validate(doc, method=None):
    """
    Keep a manually-rated Repack finished-good row internally consistent
    BEFORE core's set_basic_rate()/distribute_additional_costs() run.

    When set_basic_rate_manually is checked, core skips recalculating that
    row's basic_rate *and* basic_amount (same skipped code block) -- but it
    still later recomputes amount = basic_amount + additional_cost using
    whatever basic_amount happens to already be stored. If the user just
    edited Amount (or Rate) and basic_amount wasn't refreshed to match, that
    stale basic_amount silently overwrites their edit back to the old value.
    Sync basic_amount here so core's recompute lands on the value the user
    actually intended, however they edited the row.
    """
    if doc.stock_entry_type != "Repack":
        return

    before = doc.get_doc_before_save()
    before_items = {i.name: i for i in (before.items if before else [])}

    for item in doc.items:
        if not (item.t_warehouse and not item.s_warehouse):
            continue
        if not item.get("set_basic_rate_manually"):
            continue

        prev = before_items.get(item.name)
        if prev and flt(item.amount) != flt(prev.amount):
            # Amount was hand-edited this save -- treat it as the source of
            # truth and derive rate/basic_amount from it.
            if flt(item.qty):
                item.basic_rate = flt(item.amount) / flt(item.qty)
            item.basic_amount = flt(item.amount)
        else:
            # Nothing changed, or the user edited Rate instead -- keep
            # basic_amount in step with basic_rate.
            item.basic_amount = flt(item.qty) * flt(item.basic_rate)


def validate(doc, method):
    """
    For any stock entry where material is going out (s_warehouse is set),
    if an item row has custom_project_mandatory checked:
    1. Ensure the project field is filled
    2. Fetch the available project qty from Stock Ledger Entry
    3. Block submission if qty exceeds available project qty
    """
    _pin_rm_qty_to_work_order(doc)

    for item in doc.items:
        if not item.custom_project_mandatory:
            continue

        # Only validate items where material is going out (source warehouse is set)
        if not item.s_warehouse:
            continue

        if not item.project:
            frappe.throw(
                _("Row {0}: Project is mandatory for Item {1} as 'Project Mandatory' is checked").format(
                    item.idx, frappe.bold(item.item_code)
                )
            )

        available_qty = get_project_qty(item.item_code, item.s_warehouse, item.project)
        item.custom_project_back_qty = available_qty

        if item.qty > available_qty:
            frappe.throw(
                _("Row {0}: Qty ({1}) exceeds available Project Qty ({2}) for Item {3}, "
                  "Project {4}, Warehouse {5}").format(
                    item.idx,
                    frappe.bold(item.qty),
                    frappe.bold(available_qty),
                    frappe.bold(item.item_code),
                    frappe.bold(item.project),
                    frappe.bold(item.s_warehouse),
                )
            )

    # Calculate Total Quantity for finished goods only
    doc.total_quantity = sum(item.qty for item in doc.items if item.get("is_finished_item"))

    # ── For Repack: Match Valuation and Allow Zero Rates ──
    if doc.stock_entry_type == "Repack":
        # Items across different item groups get different Difference Accounts
        # (expense_account) from their Item Defaults. That mismatch alone posts
        # a GL entry for the account difference, on top of the real item-group
        # inventory-account movement. Force one shared Difference Account per
        # Repack so that leg only ever reflects a genuine RM/FG valuation gap.
        unify_repack_difference_account(doc)

        # ERPNext's set_basic_rate() (core, runs before this hook) recomputes an
        # incoming row's rate/amount from the outgoing cost every single save,
        # unless set_basic_rate_manually is checked on that row -- overwriting
        # any amount the user just typed. Flip it on automatically so a manual
        # edit sticks from the next save onward, without the user having to
        # find and check that box themselves.
        for item in doc.items:
            if item.t_warehouse and not item.s_warehouse:
                item.set_basic_rate_manually = 1

        if doc.get("custom_rosin_recording_reference"):
            # First: allow zero rate for outgoing items
            for item in doc.items:
                if item.s_warehouse and not item.t_warehouse:
                    item.allow_zero_valuation_rate = 1

            # Second: distribute the input value (even if 0) across outputs,
            # skipping any finished-good row the user has manually rated.
            set_repack_valuation(doc)


def _pin_rm_qty_to_work_order(doc):
    """
    For Manufacture SEs linked to a Work Order: keep every raw-material row's
    qty at the WO's planned required_qty instead of ERPNext's proportionally
    scaled value (which grows when actual FG > expected FG).
    """
    if doc.stock_entry_type != "Manufacture":
        return
    wo_name = doc.get("work_order")
    if not wo_name:
        return

    planned = {
        row.item_code: flt(row.required_qty)
        for row in frappe.get_all(
            "Work Order Item",
            filters={"parent": wo_name},
            fields=["item_code", "required_qty"],
        )
        if flt(row.required_qty) > 0
    }
    if not planned:
        return

    for item in doc.items:
        if item.is_finished_item or item.is_scrap_item:
            continue
        if not item.s_warehouse:
            continue
        pinned_qty = planned.get(item.item_code)
        if pinned_qty is None:
            continue
        if flt(item.qty) == pinned_qty:
            continue  # already correct, skip recalc
        item.qty = pinned_qty
        item.transfer_qty = flt(pinned_qty) * flt(item.conversion_factor or 1)
        item.basic_amount = flt(pinned_qty) * flt(item.basic_rate)
        item.amount = flt(pinned_qty) * flt(item.valuation_rate or item.basic_rate)


def unify_repack_difference_account(doc):
    """
    Force every item row in a Repack entry onto one shared Difference Account
    (expense_account), instead of each item's own Item Default (which varies
    by item group). With a shared account, the Difference Account GL leg nets
    to zero when RM value == FG value, and only shows a real amount when
    there's a genuine valuation gap -- matching what "Difference Account" is
    meant to represent, rather than an artifact of different item defaults.
    """
    shared_account = frappe.get_cached_value(
        "Company", doc.company, "stock_adjustment_account"
    ) or next((item.expense_account for item in doc.items if item.expense_account), None)

    if not shared_account:
        return

    for item in doc.items:
        item.expense_account = shared_account


def set_repack_valuation(doc):
    """
    Custom valuation for Rosin Repack entries.
    Formula applied per 'block' (1 Raw Material followed by N Finished Goods).
    Example block: $1,808.50 RM / 563 total FG qty = $3.21 basic rate assigned to those specific FG rows.
    """
    from frappe.utils import flt
    
    blocks = []
    current_block = None
    
    # 1. Group items into clusters (1 Raw Material -> Many Finished Goods)
    for item in doc.items:
        if item.s_warehouse and not item.t_warehouse:
            # Start of a new yield operation
            current_block = {
                "rm_item": item,
                "fg_items": []
            }
            blocks.append(current_block)
        elif item.t_warehouse and not item.s_warehouse:
            if current_block is not None:
                current_block["fg_items"].append(item)
                
    # 2. Calculate and assign valuation rate per cluster independently
    for block in blocks:
        rm_item = block["rm_item"]
        fg_items = block["fg_items"]
        
        # Calculate Raw Material Amount
        rate = flt(rm_item.valuation_rate)
        if rate <= 0:
            rate = flt(frappe.db.get_value("Stock Ledger Entry", {
                "item_code": rm_item.item_code,
                "warehouse": rm_item.s_warehouse,
                "is_cancelled": 0
            }, "valuation_rate", order_by="posting_date desc, creation desc"))
        
        rm_item.basic_rate = rate
        rm_amount = flt(rm_item.qty) * rate
        rm_item.amount = rm_amount

        # Rows the user has manually rated keep their own amount; only the
        # remaining pool (RM value minus what's manually assigned) gets
        # divided across the still-automatic finished goods.
        manual_fg = [fg for fg in fg_items if fg.get("set_basic_rate_manually")]
        auto_fg = [fg for fg in fg_items if not fg.get("set_basic_rate_manually")]
        remaining_amount = rm_amount - sum(flt(fg.amount) for fg in manual_fg)

        # Divide amount among this block's specific finished goods
        sum_fg_qty = sum(flt(fg.qty) for fg in auto_fg)
        if sum_fg_qty > 0:
            new_rate = remaining_amount / sum_fg_qty
            for fg in auto_fg:
                fg.basic_rate = flt(new_rate, 4)
                fg.amount = flt(fg.qty) * fg.basic_rate



@frappe.whitelist()
def get_project_qty(item_code, warehouse, project):
    """
    Get the available qty for an Item in a specific Warehouse and Project
    by summing actual_qty from Stock Ledger Entry.
    """
    result = frappe.db.sql(
        """
        SELECT IFNULL(SUM(actual_qty), 0) as qty
        FROM `tabStock Ledger Entry`
        WHERE item_code = %s
          AND warehouse = %s
          AND project = %s
          AND is_cancelled = 0
        """,
        (item_code, warehouse, project),
        as_dict=True,
    )

    return result[0].qty if result else 0


@frappe.whitelist()
def get_wo_rm_planned_qty(work_order):
    """Return {item_code: required_qty} for a Work Order's raw material rows."""
    rows = frappe.get_all(
        "Work Order Item",
        filters={"parent": work_order},
        fields=["item_code", "required_qty"],
    )
    return {r.item_code: flt(r.required_qty) for r in rows if flt(r.required_qty) > 0}

File: cannabis_management/public/js/stock_entry.js
frappe.ui.form.on("Stock Entry", {
    refresh: function (frm) {
        // Source/Target Tags: only offer Metric Tags whose License matches
        // the row's Source/Target Warehouse.
        cannabis_management.metric_tag.filter_by_warehouse(frm, "tags", "s_warehouse");
        cannabis_management.metric_tag.filter_by_warehouse(frm, "to_tags", "t_warehouse");
        toggle_tag_mandatory(frm);

        frm.fields_dict.items.grid.update_docfield_property("project", "reqd", 0);
        $.each(frm.doc.items || [], function (i, item) {
            if (item.custom_project_mandatory) {
                frappe.meta.get_docfield("Stock Entry Detail", "project", frm.doc.name).reqd = 1;
            }
        });
        frm.refresh_fields();
        calculate_total_quantity(frm);

        if (frm.is_new() && frm.doc.stock_entry_type === "Manufacture" && frm.doc.work_order) {
            setTimeout(() => pin_rm_qty_from_wo(frm), 120);
            _fix_operating_cost_from_wo(frm);
        }
    },

    stock_entry_type: function (frm) {
        toggle_tag_mandatory(frm);
    },

    project: function (frm) {
        if (frm.doc.project) {
            $.each(frm.doc.items || [], function (i, item) {
                frappe.model.set_value(item.doctype, item.name, "project", frm.doc.project);
                frappe.model.set_value(item.doctype, item.name, "batch", frm.doc.project);
            });
        }
    },

    items_add: function (frm, cdt, cdn) {
        if (frm.doc.project) {
            frappe.model.set_value(cdt, cdn, "project", frm.doc.project);
            frappe.model.set_value(cdt, cdn, "batch", frm.doc.project);
        }
        toggle_tag_mandatory(frm);
        calculate_total_quantity(frm);
    },

    items_remove: function (frm) {
        calculate_total_quantity(frm);
    },
});

frappe.ui.form.on("Stock Entry Detail", {
    qty: function (frm, cdt, cdn) {
        setTimeout(() => calculate_total_quantity(frm), 300);
    },

    amount: function (frm, cdt, cdn) {
        // Repack finished-good (t_warehouse only) rows: core recalculates
        // rate/amount from the outgoing cost on every save unless
        // set_basic_rate_manually is checked, and nothing client-side keeps
        // basic_rate in step with a direct Amount edit -- so without this,
        // whatever's typed here can get overwritten before/On save.
        let row = locals[cdt][cdn];
        if (frm.doc.stock_entry_type !== "Repack" || !row.t_warehouse || row.s_warehouse) {
            return;
        }
        // Conversion Entry drafts are rebalanced against the raw material cost
        // on every save, so pinning a hand-typed amount here would only be
        // overwritten -- and would freeze the row in the meantime.
        if (frm.doc.custom_conversion_entry_reference) {
            return;
        }
        if (!row.set_basic_rate_manually) {
            frappe.model.set_value(cdt, cdn, "set_basic_rate_manually", 1);
        }
        if (flt(row.qty)) {
            frappe.model.set_value(cdt, cdn, "basic_rate", flt(row.amount) / flt(row.qty));
        }
        frappe.model.set_value(cdt, cdn, "basic_amount", flt(row.amount));
    },

    is_finished_item: function (frm, cdt, cdn) {
        setTimeout(() => calculate_total_quantity(frm), 300);
    },

    item_code: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        if (row.item_code) {
            frappe.db.get_value("Item", row.item_code, "custom_project_mandatory", function (r) {
                if (r) {
                    frappe.model.set_value(cdt, cdn, "custom_project_mandatory", r.custom_project_mandatory || 0);
                }
            });
            if (frm.doc.project) {
                setTimeout(function () {
                    frappe.model.set_value(cdt, cdn, "project", frm.doc.project);
                    frappe.model.set_value(cdt, cdn, "batch", frm.doc.project);
                }, 500);
            }
        } else {
            frappe.model.set_value(cdt, cdn, "custom_project_mandatory", 0);
            frappe.model.set_value(cdt, cdn, "custom_project_back_qty", 0);
        }
    },

    custom_project_mandatory: function (frm, cdt, cdn) {
        toggle_project_mandatory(frm);
    },

    project: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        if (row.custom_project_mandatory && row.item_code && row.project && row.s_warehouse) {
            fetch_project_qty(frm, cdt, cdn, row);
        }
        if (row.project) {
            frappe.model.set_value(cdt, cdn, "batch", row.project);
        }
    },

    s_warehouse: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        if (row.custom_project_mandatory && row.item_code && row.project && row.s_warehouse) {
            fetch_project_qty(frm, cdt, cdn, row);
        }
        toggle_tag_mandatory(frm);
    },

    t_warehouse: function (frm) {
        toggle_tag_mandatory(frm);
    },
});

function toggle_project_mandatory(frm) {
    let any_mandatory = (frm.doc.items || []).some((item) => item.custom_project_mandatory);
    frm.fields_dict.items.grid.update_docfield_property("project", "reqd", any_mandatory ? 1 : 0);
    frm.refresh_fields();
}

// Repack only: Source/Target Tags become mandatory per row, based on which
// warehouse that row actually uses -- a raw-material row (Source Warehouse
// only) needs a Source Tag, a finished-good row (Target Warehouse only)
// needs a Target Tag, and never both on the same row. Outside Repack, this
// leaves the doctype's own defaults alone (Source Tags always required,
// Target Tags optional) -- only Repack rows get this per-row treatment.
// frm.set_df_property(..., row.name) scopes the change to that single grid
// row instead of the whole "tags"/"to_tags" column, which is what makes
// per-row (not per-column) mandatory possible here.
function toggle_tag_mandatory(frm) {
    let is_repack = frm.doc.stock_entry_type === "Repack";
    (frm.doc.items || []).forEach(function (row) {
        let tags_reqd = is_repack ? (row.s_warehouse ? 1 : 0) : 1;
        let to_tags_reqd = is_repack ? (row.t_warehouse ? 1 : 0) : 0;
        frm.set_df_property("items", "reqd", tags_reqd, frm.doc.name, "tags", row.name);
        frm.set_df_property("items", "reqd", to_tags_reqd, frm.doc.name, "to_tags", row.name);
    });
}

function fetch_project_qty(frm, cdt, cdn, row) {
    frappe.call({
        method: "cannabis_management.cannabis_management.custom.stock_entry.get_project_qty",
        args: {
            item_code: row.item_code,
            warehouse: row.s_warehouse,
            project: row.project,
        },
        callback: function (r) {
            if (r.message !== undefined) {
                frappe.model.set_value(cdt, cdn, "custom_project_back_qty", r.message);
            }
        },
    });
}

function pin_rm_qty_from_wo(frm) {
    frappe.call({
        method: "cannabis_management.cannabis_management.custom.stock_entry.get_wo_rm_planned_qty",
        args: { work_order: frm.doc.work_order },
        callback: function (r) {
            if (!r.message) return;
            let planned = r.message;
            let promises = [];
            (frm.doc.items || []).forEach((item) => {
                if (item.is_finished_item || item.is_scrap_item || !item.s_warehouse) return;
                let pinned = planned[item.item_code];
                if (pinned !== undefined && Math.abs(item.qty - pinned) > 0.0001) {
                    promises.push(
                        frappe.model.set_value(item.doctype, item.name, "qty", pinned)
                    );
                }
            });
            if (promises.length) {
                Promise.all(promises).then(() => frm.refresh_field("items"));
            }
        },
    });
}

function _fix_operating_cost_from_wo(frm) {
    frappe.call({
        method: "cannabis_management.doc_hooks.stock_entry.get_wo_cost_breakdown",
        args: {
            work_order: frm.doc.work_order,
            company: frm.doc.company,
            fg_completed_qty: frm.doc.fg_completed_qty,
            wo_qty: null,
        },
        callback: function (r) {
            if (!r.message || !r.message.length) return;

            // Clear all additional costs — server validate owns this field
            frm.doc.additional_costs = [];

            // Add per-account rows
            r.message.forEach(item => {
                frm.add_child("additional_costs", {
                    expense_account: item.expense_account,
                    description: item.description,
                    amount: item.amount,
                });
            });

            // Recalculate total from all rows
            let total = (frm.doc.additional_costs || []).reduce((s, c) => s + flt(c.amount), 0);
            frm.set_value("total_additional_costs", total);
            frm.refresh_field("additional_costs");
        },
    });
}

function calculate_total_quantity(frm) {
    let finished_qty = 0;
    let raw_qty = 0;

    (frm.doc.items || []).forEach((item) => {
        if (item.is_finished_item) {
            finished_qty += item.qty || 0;
        } else {
            raw_qty += item.qty || 0;
        }
    });

    const total_qty =
        frm.doc.stock_entry_type === "Repack"
            ? finished_qty
            : finished_qty + raw_qty;

    frm.doc.total_quantity = total_qty;
    frm.refresh_field("total_quantity");
}